axum/src/middleware/map_response.rs

```diff
@@ -0,0 +1,341 @@
+/// # let app: Router<_> = app;
+/// ```
+pub fn map_response_with_state<F, S, T>(state: S, f: F) -> MapResponseLayer<F, S, T> {
+    map_response_with_state_arc(Arc::new(state), f)
+}
+
+/// Create a middleware from an async function that transforms a response, with the given [`Arc`]'ed
+/// state.
+///
+/// See [`map_response_with_state`] for an example.
+///
+/// See [`State`](crate::extract::State) for more details about accessing state.
+pub fn map_response_with_state_arc<F, S, T>(state: Arc<S>, f: F) -> MapResponseLayer<F, S, T> {
+    MapResponseLayer {
+        f,
+        state,
+        _extractor: PhantomData,
+    }
+}
+
+/// A [`tower::Layer`] from an async function that transforms a response.
+///
+/// Created with [`map_response`]. See that function for more details.
+pub struct MapResponseLayer<F, S, T> {
+    f: F,
+    state: Arc<S>,
+    _extractor: PhantomData<fn() -> T>,
+}
+
+impl<F, S, T> Clone for MapResponseLayer<F, S, T>
+where
+    F: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            f: self.f.clone(),
+            state: Arc::clone(&self.state),
+            _extractor: self._extractor,
+        }
+    }
+}
+
+impl<S, I, F, T> Layer<I> for MapResponseLayer<F, S, T>
+where
+    F: Clone,
+{
+    type Service = MapResponse<F, S, I, T>;
+
+    fn layer(&self, inner: I) -> Self::Service {
+        MapResponse {
+            f: self.f.clone(),
+            state: Arc::clone(&self.state),
+            inner,
+            _extractor: PhantomData,
+        }
+    }
+}
+
+impl<F, S, T> fmt::Debug for MapResponseLayer<F, S, T>
+where
+    S: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("MapResponseLayer")
+            // Write out the type name, without quoting it as `&type_name::<F>()` would
+            .field("f", &format_args!("{}", type_name::<F>()))
+            .field("state", &self.state)
+            .finish()
+    }
+}
+
+/// A middleware created from an async function that transforms a response.
+///
+/// Created with [`map_response`]. See that function for more details.
+pub struct MapResponse<F, S, I, T> {
+    f: F,
+    inner: I,
+    state: Arc<S>,
+    _extractor: PhantomData<fn() -> T>,
+}
+
+impl<F, S, I, T> Clone for MapResponse<F, S, I, T>
+where
+    F: Clone,
+    I: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            f: self.f.clone(),
+            inner: self.inner.clone(),
+            state: Arc::clone(&self.state),
+            _extractor: self._extractor,
+        }
+    }
+}
+
+macro_rules! impl_service {
+    (
+        $($ty:ident),*
+    ) => {
+        #[allow(non_snake_case, unused_mut)]
+        impl<F, Fut, S, I, B, ResBody, $($ty,)*> Service<Request<B>> for MapResponse<F, S, I, ($($ty,)*)>
+        where
+            F: FnMut($($ty,)* Response<ResBody>) -> Fut + Clone + Send + 'static,
+            $( $ty: FromRequestParts<S> + Send, )*
```

Comment:
Note that we require `FromRequestParts` and not `FromRequest`. Not sure there would be a way to recover the request afterwards so we can pass it to the inner service 🤔

Reply:
Yeah, this is exactly right.